### PR TITLE
Backport of [NET-6138] security: Bump google.golang.org/grpc to 1.56.3 (CVE-2023-44487) to release/1.15.x

### PR DESCRIPTION
**Pull request by zalimeni**
## Backport

This PR is manually generated from #19414 to be assessed for backporting.

This backport also includes various fixes from later release versions required for
tests and linters to pass. See 77f44fa87823ac59cbe41068b742b9e02c122ed8 for the majority of these changes.

The below text is copied from the body of the original PR.

---

### Description

Upgrade `google.golang.org/grpc` to 1.56.3 (where 1.57.x is already in use, upgrade to 1.57.2) to mitigate [CVE-2023-44487](https://nvd.nist.gov/vuln/detail/CVE-2023-44487).

1.56.3 was chosen as the earliest minor release line to receive a patch for this vulnerability in order to minimize change prior to the 1.17 release of Consul; a later upgrade should consider moving to the latest minor version of gRPC. A review of the release notes and [changelog](https://github.com/grpc/grpc-go/compare/v1.55.0...v1.56.3) between 1.55.0 and 1.56.3 was conducted to check for unreported breaking changes.

### Testing & Reproduction steps

CI tests should continue to pass.

### Links

* [CVE-2023-44487](https://nvd.nist.gov/vuln/detail/CVE-2023-44487)

### PR Checklist

* [ ] updated test coverage
* [ ] external facing docs updated
* [x] appropriate backport labels added
* [x] not a security concern

### CODE REVIEW

File: agent/xds/z_xds_packages.go [hunk 1]
```diff
@@ -3,16 +3,32 @@
 package xds
```
Comment (zalimeni) [inline agent/xds/z_xds_packages.go:3]: @cthain do the changes here and in other non-mod files (`troubleshoot/proxy/z_xds_packages.go` and `build-support/scripts/envoy-library-references.sh`) still look correct to you, based on your compatibility fixes?

File: proto-public/go.mod [hunk 2]
```diff
@@ -4,18 +4,18 @@ go 1.19
 
 require (
 	github.com/stretchr/testify v1.5.1
-	google.golang.org/grpc v1.37.1
-	google.golang.org/protobuf v1.27.1
+	google.golang.org/grpc v1.56.3
+	google.golang.org/protobuf v1.30.0
```
Comment (zalimeni) [inline proto-public/go.mod:8]: Note that this version was _significantly_ out of date compared w/ the root `go.mod`.

**Comment by zalimeni** [comment]
Looks like we may have some incompatibilities for this one: https://github.com/hashicorp/consul/actions/runs/6672184414/job/18135656338?pr=19421#step:6:24

**Review by cthain — APPROVED** [review]
:+1:

File: agent/grpc-internal/resolver/resolver.go [hunk 3]
```diff
@@ -93,7 +93,7 @@ func (s *ServerResolverBuilder) Build(target resolver.Target, cc resolver.Client
 	}
 
 	//nolint:staticcheck
-	serverType, datacenter, err := parseEndpoint(target.Endpoint)
+	serverType, datacenter, err := parseEndpoint(target.Endpoint())
```
Comment (zalimeni) [inline agent/grpc-internal/resolver/resolver.go:96]: This and below one-line changes fix binary incompatibilities introduced by the upgrade; these match later release branches and `main`

File: build-support/scripts/envoy-library-references.sh [hunk 4]
```diff
@@ -28,7 +28,7 @@ if [[ ! -f GNUmakefile ]] || [[ ! -f go.mod ]]; then
     exit 1
```
Comment (zalimeni) [inline build-support/scripts/envoy-library-references.sh:28]: This similarly now matches `main` and needed to change to support deps upgrades

File: proto-public/pbacl/acl.pb.go [hunk 5]
```diff
@@ -1,6 +1,6 @@
 // Code generated by protoc-gen-go. DO NOT EDIT.
 // versions:
-// 	protoc-gen-go v1.28.1
+// 	protoc-gen-go v1.31.0
```
Comment (zalimeni) [inline proto-public/pbacl/acl.pb.go:3]: Here and below: all comment changes generated by `make proto` that came with upgrade, necessary to satisfy linter (and needed bc we changed the `proto-gen-go` version)

**Review by david-yu — APPROVED** [review]

**Review by hashi-derek — APPROVED** [review]

File: agent/grpc-internal/resolver/resolver_test.go [hunk 6]
```diff
@@ -40,7 +41,7 @@ func TestServerResolverBuilder(t *testing.T) {
 		_, err := rs.Build(resolver.Target{
 			Scheme:    "consul",
 			Authority: rs.Authority(),
-			Endpoint:  endpoint,
+			URL:       url.URL{Opaque: endpoint},
```
Comment (zalimeni) [inline agent/grpc-internal/resolver/resolver_test.go:44]: See https://github.com/hashicorp/consul/commit/77f44fa87823ac59cbe41068b742b9e02c122ed8

File: test/integration/connect/envoy/test-sds-server/sds.go [hunk 7]
```diff
@@ -140,7 +140,7 @@ func makeLoggerCallbacks(log hclog.Logger) *xds.CallbackFuncs {
 			log.Trace("gRPC stream opened", "id", id, "addr", addr)
```
Comment (zalimeni) [inline test/integration/connect/envoy/test-sds-server/sds.go:140]: See https://github.com/hashicorp/consul/commit/77f44fa87823ac59cbe41068b742b9e02c122ed8

**Pull request merged**